datafusion/common/src/config.rs

```diff
@@ -188,6 +195,211 @@ macro_rules! config_namespace {
     }
 }
 
+#[derive(Serialize, Deserialize, Debug)]
+pub struct EncryptionKeyAsHex {
+    pub key: String,
+    pub key_metadata: Option<String>,
+}
+
+#[derive(Debug)]
+pub struct EncryptionKey {
+    pub key: Vec<u8>,
+    pub key_metadata: Option<Vec<u8>>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct EncryptionColumnKeys {
+    pub column_keys_as_hex: HashMap<String, EncryptionKeyAsHex>,
+}
+
+impl EncryptionColumnKeys {
+    pub fn new(
+        column_names: &Vec<String>,
+        keys: &Vec<Vec<u8>>,
+        meta: &Vec<Vec<u8>>,
+    ) -> Self {
+        let mut column_keys_as_hex: HashMap<String, EncryptionKeyAsHex> = HashMap::new();
+
+        for (i, column_name) in column_names.iter().enumerate() {
+            let key_metadata: Option<String> = meta.get(i).map(|x| hex::encode(x));
+            let encryption_key = EncryptionKeyAsHex {
+                key: hex::encode(keys[i].clone()),
+                key_metadata,
+            };
+            column_keys_as_hex.insert(column_name.clone(), encryption_key);
+        }
+
+        EncryptionColumnKeys { column_keys_as_hex }
+    }
+
+    pub fn to_json(&self) -> String {
+        serde_json::to_string(self).unwrap()
+    }
+
+    pub fn from_json_to_column_keys(s: &String) -> HashMap<String, EncryptionKey> {
+        let eck: EncryptionColumnKeys =
+            serde_json::from_str(s).expect("failed to decode column keys from JSON");
+        let mut hm: HashMap<String, EncryptionKey> = HashMap::new();
+        for (key, val) in eck.column_keys_as_hex {
+            let col = key;
+            let key = hex::decode(val.key).expect("Invalid column key");
+            let key_metadata = val
+                .key_metadata
+                .map(|x| hex::decode(x).expect("Invalid column metadata"));
+            let encryption_key = EncryptionKey { key, key_metadata };
+            hm.insert(col, encryption_key);
+        }
+        hm
+    }
+}
+
+config_namespace! {
+    pub struct ConfigFileEncryptionProperties {
+        pub encrypt_footer: bool, default = false
+        pub footer_key_as_hex: String, default = String::new()
+        pub footer_key_metadata_as_hex: String, default = String::new()
+        pub column_keys_as_json_hex: String, default = String::new()
```

Comment:
Rather than encoding the column keys as JSON, I wonder if it would be better to follow the existing approach for column specific options in `TableParquetOptions` where they use a HashMap of options and use a `::` separator between the column name and config field name. That does mean having to manually implement the `ConfigField` trait rather than being able to use the `config_namespace` macro, but it doesn't look too complicated (eg. see the [TableParquetOptions impl](https://github.com/corwinjoy/datafusion/blob/5a2b4562f9d6831fe6629296cdc151e3e9e582f7/datafusion/common/src/config.rs#L1958))

Reply:
One issue is that columns need both a key and potentially metadata instead of a single value so I'm not quite sure how that part would work. Thoughts?

Reply:
I think this should still work fine if the value is a struct rather than a plain string option, eg. see `column_specific_options` which works a bit like this. It looks like there's an implementation of `ConfigField` for hash maps [here](https://github.com/corwinjoy/datafusion/blob/21248fbbb4db9f30a3bbc92561b29a71ab0992bd/datafusion/common/src/config.rs#L1837) that is used.

If I've understood this right, the string paths for the config options will end up looking something like `format.file_decryption_properties.key::column_name` and `format.file_decryption_properties.key_metadata::column_name` .

Reply:
@adamreeve Well, I was able to add this but there are a couple problems.
1. The code is a bit lengthly and ugly which makes it somewhat harder to maintain since we can no longer use the macros.
2. I could not get the "::" syntax to work. There is some kind of bug in the macro code where it won't properly recognize fields containing "::" which generates an error before I can even get to set. It's something weird in the macro code that I don't really understand. So I had to go with another "." for now.

Reply:
Thanks, that explanation makes sense. Let's talk about the crypto namespace in our catchup today. I have a bad feeling about doing that / worry that it might not be compatible with a lot of datafusion operations.